rclcpp/include/rclcpp/any_subscription_callback.hpp

```diff
@@ -176,6 +177,26 @@ class AnySubscriptionCallback
     }
   }
 
+  void dispatch_intra_process(
+    ConstMessageSharedPtr & message, const rmw_message_info_t & message_info)
+  {
+    (void)message_info;
+    if (const_shared_ptr_callback_) {
+      const_shared_ptr_callback_(std::move(message));
```

Comment:
The shared_ptr passed here is taken from the IPM, which gives you a copy of the shared_ptr stored in the buffer or the ownership of the stored message in case all the other subscription has been handled. So it's fine to move it, as it's not used again.

It's true that this is a little confusing, I will try to do it in a cleaner way.

Reply:
I just wonder if the move is necessary here, if it boils down to essentially:

```c++
std::shared_ptr<T> message = func();
const_shared_ptr_callback_(message);
```

Then the compiler will automatically use move semantics because it knows `message` is not used after being passed to the callback anyways.

At least that's what I think will happen, we could try the `-Wredundant-move` warning to see if the compiler will tell us, see: https://vmpstr.blogspot.com/2015/12/redundant-stdmove.html

Reply:
I installed clang and built with `--Wredundant-move` and `-Wpessimizing-move`, no new warning appeared.
Still, I will change the signature to `ConstMessageSharedPtr`, stop using `std::move` explicitly, and let the compiler do it efficiently. That's maybe the standard way. To have a consistent style, I will do the same in the old method with `unique_ptr` signature:

https://github.com/ros2/rclcpp/blob/d5dc237099262565622d13a8203ffb0f8367e337/rclcpp/include/rclcpp/any_subscription_callback.hpp#L200-L220

I will apply similar changes in other methods with similar signatures that were modified or added in this PR.